images/devtools-golang-v1beta1/context/Makefile

```diff
@@ -54,146 +58,234 @@ endef
+## Variable APP_DOCKERFILE : The path to the Dockerfile to use for building OCI images. Defaults to `build/package/Dockerfile`.
+## Variable BUILD_OCI : boolean (true,false) indicating if OCI should be built. Defaults to true if $(APP_DOCKERFILE) exists
+## Variable ENABLE_BUF: boolean (true,false) indicating if buf should be enabled. Defaults to true if buf.work.yaml exists
 APP_EXECUTABLE?=$(APP_NAME)
 APP_DOCKERFILE?=build/package/Dockerfile
 BUILD_OCI?=$(if $(wildcard $(APP_DOCKERFILE)),true,false)
 XDG_CACHE_HOME?=$(HOME)/.cache
+ENABLE_BUF?=$(if $(wildcard buf.work.yaml),true,false)
 
 # setting this so that it is more likely things will work on MacOS and Linux
 export POSIXLY_CORRECT:=1
 
-golangci_lint=golangci-lint $(if $(filter all commands,$(VERBOSE)),-v)
+# Get git information
 giti_description=$(shell git describe --tags --always)
 giti_commit_hash=$(shell git log -1 --format="%H")
 giti_commit_date=$(shell git log -1 --format="%aI")
 giti_url:=$(shell git remote get-url origin)
+
+# Define golang flags
 go_mod:=$(shell go list -m)
-go_ldflags= \
+go_ldflags?= \
 	-ldflags \
 	"-X '$(go_mod)/internal/version.gitCommit=$(giti_commit_hash)' \
 	-X '$(go_mod)/internal/version.gitRemote=$(giti_url)' \
 	-X '$(go_mod)/internal/version.gitDesc=$(giti_description)' \
 	-X '$(go_mod)/internal/version.gitCommitDate=$(giti_commit_date)'" \
 
-go_build_flags=
+go_build_flags?=
 
-stamps_dir=var/stamps
-outputs_dir=var/outputs
+localstatedir?=var
+generated_dir?=$(localstatedir)/generated
+stamps_dir?=$(localstatedir)/stamps
+outputs_dir?=$(localstatedir)/outputs
 
 $(call dump_vars,APP_EXECUTABLE APP_DOCKERFILE BUILD_OCI XDG_CACHE_HOME)
 
 ########################################################################
-# targets ...
+# generic targets
 ########################################################################
 
 .PHONY: clean
-clean: ## clean build outputs
-clean: clean-$(outputs_dir)/
-	rm -vr coverage.out || :
-	go clean
+clean: ## Clean outputs.
+clean: clean-$(localstatedir)/
 
-.PHONY: validate
-validate: validate-static test  ## Run static and dynamic validation (tests)
+.PHONY: configure
+configure: ## Configure the project.
+
+.PHONY: generate
+generate: ## Generate all intermediate code.
 
 .PHONY: validate-fix
-validate-fix:  ## Let the tooling fix what can be automatically fixed
+validate-fix: ## Fix auto-fixable validation errors.
+
 .PHONY: validate-static
-validate-static:  ## Static validation
+validate-static: ## Run static validation.
+
 .PHONY: test
-test:  ## Dynamic code validation
+test: ## Run the project's tests.
+
+.PHONY: validate
+validate: ## Validate everything.
+validate: validate-static test
+
+.PHONY: build
+build: ## Build the source.
+
+.PHONY: watch
+watch: ## Watch code for changes and re-validate.
+
+.PHONY: install
+install: ## Install globally.
+
 .PHONY: publish
-publish: ## publish artifacts
+publish: ## Publish artifacts.
+
+.PHONY: deploy
+deploy: ## Deploy the project.
+
+.PHONY: all
```

Comment:
duplicate command from line 12?

Reply:
It is, however I think it is better to duplicate it, as it makes things look consistent and `.PHONY: all` can be repeated and has no ill effect.

The reason it is at the top of the file is because the `all` target must be defined before any other targets, I could maybe move it up inside the generic targets section but if someone defines a target in `devtools-settings.mk` then it will not work as expected.

Also, I want to have all generic targets documented in one place with their actual dependencies, and I want them all to look more or less consistent. I want it to be clear when looking at this section all the targets here are `.PHONY`.

I could maybe remove the first `.PHONY`, I think things may still work the same if I do, but I'm not entirely convinced as every example in the documentation has `.PHONY` preceding the actual target.

Reply:
Will look at it in a separate PR if someone disagrees.